design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+- Outputs from one pass as inputs to another
+- Optimize out passes without geometry
+- Expose all of this outside the toolkit
+
+#### Benefits
+Similar to what was discussed above:
+- Using the rasterizer to sort out your data ahead of time.  We do this a lot with weather data.
+- Reducing the size of a very costly shader.  The bigger your shader, the longer it takes.  If you have some fancy effect, it can probably be rendered at lower resolution and then laid over the top of the final scene.
+- Doing post-process effects on the map.  In games you might think of lens warping, wipes, and such.  For mapping, this tends to be more merging data layers or things like glow.
+
+### Off Screen Render Targets
+
+_Addresses core functionality [#5](#core)._
+
+The toolkit already has support for one offscreen render target.  You can either set things up “normally” where the toolkit draws to a window or you can set things up to draw to memory and it will hand you back the resulting image.  This is a little like that, but in real time.
+
+Having an explicit off screen render target lets the developer do something like a snapshot in real time.  That is, run some sort of logic on the data being rendered before it gets its final visual representation.
+
+#### Required Changes:
+To support Rendering Passes we’ll need Render Targets we can access in a way similar to a texture.  They’ll be set up with:
+- Settings such as bit depth and width/height (you don’t always want full screen resolution)
+- Allowing geometry to specify which one(s) they’ll render to as a “target”
+- Query and snapshot
+
+This may seem a bit of a detour, but it’s fairly easy to get the basics of it in place and allow for further development as developers might add new shaders and varieties of targets.
+
+#### Benefits
+It's a core idea in many rendering toolkits, but for MapLibre Native it will let us:
+- Reach into the rendering pipeline to make use of processed data without adding a GPU<->CPU round trip.
+- Make it easier to assemble things like videos in real time.
+- Help define how the rendering passes work with fixed points of output.
+
+### Snapshotting
+
+When you take a snapshot of the map, the toolkit will return an image to you.  This works well for OpenGL, but not as well for other SDKs.  What we really need to do is provide a callback and copy the buffer after it's finished drawing.
+
+For OpenGL this will be about the same.  For other SDKs this can be much more efficient.
+
+#### Required Changes:
+
+The snapshot method will need to take a callback and notify it when a snapshot is ready.
+We may also want to expand this out to get multiple snapshots, or at least allow for that in the future.
+
+#### Benefits
+
+For non-GL SDKs:
+- The caller won't be interrupting the rendering flow and forcing it to come to a stop.
+- If implemented efficiently, the caller could get multiple frames at very high speed.  Perhaps high enough speed to encode video.
+
+### Drawables and Builders
```

Comment:
It appears to me drawables may serve as an effective replacement for the existing render tree? Many facets of the render tree are less than optimal, ie the annotation manager having a single mutex for the whole object the tree has to lock. Other bits like the glyph manager also appear to have been built with no consideration for parallel access.

How far to you plan to take this drawable concept, with respect to the above?

Reply:
Drawables map a little closer to Layers & Buckets, but not exactly.  The RenderTree I'd call more of a "Scene", though again not exactly.  I do think it might need to be made more general.  

Rather than addressing that explicitly, I think I'd let the developers get Drawables working and then see how far they want to push the concept.  So the RenderTree might change a bit, but probably not get completely replaced.